kafka/src/main/java/org/apache/apex/malhar/kafka/AbstractExactlyOnceKafkaOutputOperator.java

```diff
@@ -0,0 +1,366 @@
+      return;
+    }
+
+    List<TopicPartition> topicPartitions = new ArrayList<>();
+
+    for (Map.Entry<Integer,Long> entry: currentOffsets.entrySet()) {
+
+      topicPartitions.add(new TopicPartition(getTopic(), entry.getKey()));
+    }
+
+    consumer.assign(topicPartitions);
+
+    for (Map.Entry<Integer,Long> entry: currentOffsets.entrySet()) {
+
+      Long storedOffset = 0L;
+      Integer currentPartition = entry.getKey();
+      Long currentOffset = entry.getValue();
+
+      if (storedOffsets.containsKey(currentPartition)) {
+        storedOffset = storedOffsets.get(currentPartition);
+      }
+
+      if (storedOffset >= currentOffset) {
+        continue;
+      }
+
+      consumer.seek(new TopicPartition(getTopic(), currentPartition), storedOffset);
+
+      int kafkaAttempt = 0;
+
+      while ( true ) {
+
+        ConsumerRecords<String, String> consumerRecords = consumer.poll(100);
+
+        if (consumerRecords.count() == 0) {
+          if (kafkaAttempt++ == KAFKA_CONNECT_ATTEMPT) {
+            break;
+          }
+        } else {
+          kafkaAttempt = 0;
+        }
+
+        boolean crossedBoundary = false;
+
+        for (ConsumerRecord consumerRecord : consumerRecords) {
+
+          if (!doesKeyBelongsToThisInstance(operatorId, (String)consumerRecord.key())) {
+            continue;
+          }
+
+          T value = (T)consumerRecord.value();
+
+          if ( partialWindowTuples.containsKey(value)) {
+            Integer count = partialWindowTuples.get(value);
+            partialWindowTuples.put(value, count + 1);
+          } else {
+            partialWindowTuples.put(value, 1);
+          }
+
+          if (consumerRecord.offset() >= currentOffset) {
+            crossedBoundary = true;
+            break;
+          }
+        }
+
+        if (crossedBoundary) {
+          break;
+        }
+      }
+    }
+  }
+
+  private Map<Integer,Long> getCurrentOffsets()
+  {
+    Map<Integer, Long> currentOffsets;
+
+    try {
+      currentOffsets = getPartitionsAndOffsets();
+    } catch (ExecutionException e) {
+      throw new RuntimeException(e);
+    } catch (InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+
+    return currentOffsets;
+  }
+
+  private Map<Integer,Long> getStoredOffsets()
+  {
+    Map<Integer,Long> storedOffsets;
+
+    try {
+      storedOffsets = (Map<Integer,Long>)this.windowDataManager.load(operatorId, windowId);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    return storedOffsets;
+  }
+
+  private KafkaConsumer KafkaConsumerInit()
+  {
+    Properties props = new Properties();
+
+    props.put(BOOTSTRAP_SERVERS_CONFIG, getProperties().get(BOOTSTRAP_SERVERS_CONFIG));
+    props.put(KEY_DESERIALIZER_CLASS_CONFIG, KEY_SERIALIZER);
+    props.put(VALUE_DESERIALIZER_CLASS_CONFIG, VALUE_SERIALIZER);
+
+    return new KafkaConsumer<>(props);
```

Comment:
producer connection seems to be open all the time.Can you add connections.max.idle.ms as a property here so the idle connections are closed after the given time limit.